services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py

```diff
@@ -94,58 +95,98 @@ async def pull_images(
+# {'status': 'Digest: sha256:27cb6e6ccef575a4698b66f5de06c7ecd61589132d5a91d098f7f3f9285415a9'}
+# {'status': 'Status: Downloaded newer image for ubuntu:latest'}
+
+
+class _TargetPullStatus(str, Enum):
+    # They contain 'progressDetail'
+    DOWNLOADING = "Downloading"
+    DOWNLOAD_COMPLETE = "Download complete"
+    EXTRACTING = "Extracting"
+    PULL_COMPLETE = "Pull complete"
 
 
 def _parse_docker_pull_progress(
-    docker_pull_progress: dict[str, Any],
-    image_name: str,
-    all_image_pulling_data: _LayerInfoDict,
+    docker_pull_progress: _DockerProgressDict, image_pulling_data: _LayersInfoDict
 ) -> bool:
-    if docker_pull_progress.get("status") in _PULL_PROGRESS_STATES:
-        layer_id = docker_pull_progress["id"]
 
-        if docker_pull_progress["status"] == "Downloading":
-            all_image_pulling_data[image_name][layer_id] = (
+    status: Optional[str] = docker_pull_progress.get("status")
+
+    if status in list(_TargetPullStatus):
+        layer_id: LayerId = docker_pull_progress["id"]
+        # inits (read/write order is not guaranteed)
+        image_pulling_data.setdefault(layer_id, (0, 0))
+
+        if status == _TargetPullStatus.DOWNLOADING:
+            # writes
+            image_pulling_data[layer_id] = (
                 round(
                     _DOWNLOAD_RATIO * docker_pull_progress["progressDetail"]["current"]
                 ),
                 docker_pull_progress["progressDetail"]["total"],
             )
-        elif docker_pull_progress["status"] == "Downloading complete":
-            _, layer_total_size = all_image_pulling_data[image_name][layer_id]
-            all_image_pulling_data[image_name][layer_id] = (
+        elif status == _TargetPullStatus.DOWNLOAD_COMPLETE:
+            # reads
+            _, layer_total_size = image_pulling_data[layer_id]
+            # writes
+            image_pulling_data[layer_id] = (
                 round(_DOWNLOAD_RATIO * layer_total_size),
                 layer_total_size,
             )
-        elif docker_pull_progress["status"] == "Extracting":
-            _, layer_total_size = all_image_pulling_data[image_name][layer_id]
-            all_image_pulling_data[image_name][layer_id] = (
+        elif status == _TargetPullStatus.EXTRACTING:
+            # reads
+            _, layer_total_size = image_pulling_data[layer_id]
+
+            # writes
+            image_pulling_data[layer_id] = (
                 round(
                     _DOWNLOAD_RATIO * layer_total_size
                     + (1 - _DOWNLOAD_RATIO)
                     * docker_pull_progress["progressDetail"]["current"]
                 ),
                 layer_total_size,
             )
-        elif docker_pull_progress["status"] == "Pull complete":
-            _, layer_total_size = all_image_pulling_data[image_name][layer_id]
-            all_image_pulling_data[image_name][layer_id] = (
+        elif status == _TargetPullStatus.PULL_COMPLETE:
```

Comment:
ok not sure I get the whole issue. so sometimes the pull_complete or download_complete or extracting would come without downloading? is that the issue?

Reply:
the idea is that in some cases, the code READS first and in other WRITES to ``all_image_pulling_data``. But if it happens to read before it writes, then it raises a KeyError